src/multivariate/mvhypergeometric.jl

```diff
@@ -0,0 +1,125 @@
+generalizes the *hypergeometric distribution*. Consider ``n`` draws from a finite population containing ``k`` types of elements. Suppose that the population has size ``M`` and there are ``m_i`` elements of type ``i`` for ``i = 1, .., k`` with ``m_1+...m_k = M``. Let ``X = (X_1, ..., X_k)`` where ``X_i`` represents the number of elements of type ``i`` drawn, then the distribution of ``X`` is a multivariate hypergeometric distribution. Each sample of a multivariate hypergeometric distribution is a ``k``-dimensional integer vector that sums to ``n`` and satisfies ``0 \\le X_i \\le m_i``. 
+
+
+The probability mass function is given by
+
+```math
+f(x; m, n) = {{{m_1 \\choose x_1}{m_2 \\choose x_2}\\cdots {m_k \\choose x_k}}\\over {N \\choose n}}, 
+\\quad x_1 + \\cdots + x_k = n, \\quad x_i \\le m_i
+```
+
+```julia
+MvHypergeometric(m, n)   # Multivariate hypergeometric distribution for a population with
+                         # m = (m_1, ..., m_k) elements of type 1 to k and n draws
+```
+"""
+struct MvHypergeometric <: DiscreteMultivariateDistribution
+    m::AbstractVector{Int}  # number of elements of each type
+    n::Int                  # number of draws
+    function MvHypergeometric(m::AbstractVector{Int}, n::Int; check_args::Bool=true)
+        @check_args(
+            MvHypergeometric,
+            (m, all(m .>= zero.(n))),
+            zero(n) <= n <= sum(m),
+        )
+        new(m, n)
+    end
+end
+
+
+# Parameters
+
+ncategories(d::MvHypergeometric) = length(d.m)
+length(d::MvHypergeometric) = ncategories(d)
+nelements(d::MvHypergeometric) = d.m
+ntrials(d::MvHypergeometric) = d.n
+
+params(d::MvHypergeometric) = (d.m, d.n)
+
+# Statistics
+
+mean(d::MvHypergeometric) = d.n .* d.m ./ sum(d.m)
+
+function var(d::MvHypergeometric)
+    m = nelements(d)
+    k = length(m)
+    n = ntrials(d)
+    M = sum(m)
+    p = m / M
+    f = n * (M - n) / (M - 1)
+
+    v = Vector{Real}(undef, k)
+    for i = 1:k
+        @inbounds p_i = p[i]
+        v[i] = f * p_i * (1 - p_i)
+    end
+    v
+end
+
+function cov(d::MvHypergeometric)
+    m = nelements(d)
+    k = length(m)
+    n = ntrials(d)
+    M = sum(m)
+    p = m / M
+    f = n * (M - n) / (M - 1)
+
+    C = Matrix{Real}(undef, k, k)
+    for j = 1:k
+        pj = p[j]
+        for i = 1:j-1
+            @inbounds C[i, j] = -f * p[i] * pj
+        end
+
+        @inbounds C[j, j] = f * pj * (1 - pj)
+    end
+
+    for j = 1:k-1
+        for i = j+1:k
+            @inbounds C[i, j] = C[j, i]
+        end
+    end
+    C
+end
+
+
+# Evaluation
+function insupport(d::MvHypergeometric, x::AbstractVector{T}) where T<:Real
+    k = length(d)
+    m = nelements(d)
+    length(x) == k || return false
+    s = 0.0
+    for i = 1:k
+        @inbounds xi = x[i]
+        if !(isinteger(xi) && xi >= 0 && xi <= m[i])
+            return false
+        end
+        s += xi
+    end
+    return s == ntrials(d)  # integer computation would not yield truncation errors
+end
```

Comment:
Again, use of `@inbounds` is unsafe here. Moreover, I don't see why `s` should be initialized as `Float64`.

Maybe use something like
```suggestion
function insupport(d::MvHypergeometric, x::AbstractVector{<:Real})
    return length(x) == length(d) && (eltype(x) <: Integer || all(isinteger, x)) && all(((xi, mi),) -> zero(xi) <= xi <= mi, zip(x, d.m)) && sum(x) == ntrials(d)
end
```